##Question 1

Consider the following Python module:



```
a = 0
def  b():
global  a
a  = c(a)  

def  c(a):
return  a  +  2  

After importing the module into the interpreter, you execute:  
>>> b()
>>> b()
>>> b()
>>> a ?
```

What value is displayed when the last expression (a) is evaluated? Explain your answer by indicating what happens in every executed statement.

##Answer:

- When the last expression a is evaluated after executing the statements b(), b(), and b(), the value displayed would be 6.

- Initially, the variable a is assigned the value of 0 outside of any function.
- The function b() is defined. Inside b(), the global keyword is used to indicate that the variable a being referenced is the global variable a defined outside the function. Within b(), the line a = c(a) is executed. This line calls the function c() with the current value of a (which is 0). The function c(a) returns a + 2, so c(0) returns 0 + 2 = 2. The returned value is then assigned back to the global variable a, updating its value to 2.
- The third call to b() is made with b(). Inside b(), the value of a is now 6. The same steps are executed, but this time, the updated value of a is not assigned to any variable or returned. Therefore, the value of a remains 6.

##Question 2

Function fileLength(), given to you, takes the name of a file as input and returns the length of the file:  

>>>  fileLength('midterm.py') 284

>>>  fileLength('idterm.py')



```
Traceback  (most  recent  call  last): File  "<pyshell#34>",  line  1,  in  <module> fileLength('idterm.py') File  "/Users/me/midterm.py",  line  3,  in  fileLength infile  =  open(filename) FileNotFoundError:  [Errno  2]  No  such  file  or  directory: 'idterm.py'
```

  - As shown above, if the file cannot be found by the interpreter or if it cannot be read as a text file, an exception will be raised. Modify function fileLength() so that a friendly message is printed instead:  
  
  >>>  fileLength('midterm.py') 358
  
  >>>  fileLength('idterm.py') File  idterm.py  not  found.

In [1]:
def file_length(file_name):
  try:
    file = open(file_name)
    contents = file.read()
    file.close()
    print(len(contents))
  except FileNotFoundError:
    print("File", file_name, "not found.")

In [2]:
file_length("/content/Filelength(1).txt")

123


In [3]:
file_length("/content/Filelength.txt")

File /content/Filelength.txt not found.


##Question 3

- Write a class named Marsupial that can be used as shown below:  

>>>  m  =  Marsupial()

>>>  m.put_in_pouch('doll')

>>>  m.put_in_pouch('firetruck')

>>>  m.put_in_pouch('kitten')

>>>  m.pouch_contents()  ['doll',  'firetruck',  'kitten']

- Now write a class named Kangaroo as a subclass of Marsupial that inherits all the attributes of Marsupial and also: a. extends the Marsupial  __init__ constructor to take, as input, the coordinates x and y of the Kangaroo object, b. supports method jump that takes number values dx and dy as input and moves the kangaroo by dx units along the x-axis and by dy units along the y- axis, and c. overloads the __str__ operator so it behaves as shown below.  

>>>  k  =  Kangaroo(0,0)

>>>  print(k) I  am  a  Kangaroo  located  at  coordinates  (0,0)

>>>  k.put_in_pouch('doll')

>>>  k.put_in_pouch('firetruck')

>>>  k.put_in_pouch('kitten')

>>>  k.pouch_contents()  ['doll',  'firetruck',  'kitten']

>>>  k.jump(1,0)

>>>  k.jump(1,0)

>>>  k.jump(1,0)

>>>  print(k) I  am  a  Kangaroo  located  at  coordinates  (3,0)

In [4]:
class Marsupial:
    def __init__(self):
        self.pouch = []

    def put_in_pouch(self, item):
        self.pouch.append(item)

    def pouch_contents(self):
        return self.pouch


class Kangaroo(Marsupial):
    def __init__(self, x, y):
        super().__init__()
        self.x = x
        self.y = y

    def jump(self, dx, dy):
        self.x += dx
        self.y += dy

    def __str__(self):
        return "I am a Kangaroo located at coordinates ({},{})".format(self.x, self.y)


In [6]:
m = Marsupial()
m.put_in_pouch('doll')
m.put_in_pouch('firetruck')
m.put_in_pouch('kitten')
print(m.pouch_contents())

k = Kangaroo(0, 0)
print(k)
k.put_in_pouch('doll')
k.put_in_pouch('firetruck')
k.put_in_pouch('kitten')
print(k.pouch_contents())
k.jump(1, 0)
k.jump(1, 0)
k.jump(1, 0)
print(k)

['doll', 'firetruck', 'kitten']
I am a Kangaroo located at coordinates (0,0)
['doll', 'firetruck', 'kitten']
I am a Kangaroo located at coordinates (3,0)


##Question 4

- Write function collatz() that takes a positive integer x as input and prints the Collatz sequence starting at x. A Collatz sequence is obtained by repeatedly applying this rule to the previous number x in the sequence:  

x = {  𝑥/2 𝑖𝑓 𝑥 𝑖𝑠 𝑒𝑣𝑒𝑛 3𝑥

x = {  + 1 𝑖𝑓 𝑥 𝑖𝑠 𝑜𝑑𝑑  

- Your function should stop when the sequence gets to number 1. Your implementation must be recursive, without any loops.

>>>  collatz(1)

1

>>>  collatz(10)

10
5
16
8
4
2
1

In [7]:
def collatz(x):
    print(x)
    if x == 1:
        return
    elif x % 2 == 0:
        collatz(x // 2)
    else:
        collatz(3 * x + 1)

In [9]:
collatz(1)
collatz(10)

1
10
5
16
8
4
2
1


##Question 5

Write a recursive method binary() that takes a non-negative integer n and prints the binary representation of integer n.  

>>  binary(0) 0

>>  binary(1) 1

>>  binary(3) 11

>>  binary(9) 1001

In [39]:
def binary(n):
    if n > 1:
        binary(n // 2)
    print(n % 2, end=' ')


In [40]:
A = binary(0)  # Output: 0
B = binary(1)  # Output: 1
C = binary(3)  # Output: 11
D = binary(9)  # Output: 1001

0 1 1 1 1 0 0 1 

##Question 6

- Implement a class named HeadingParser that can be used to parse an HTML document, and retrieve and print all the headings in the document. You should implement your class as a subclass of HTMLParser, defined in Standard Library module html.parser. When fed a string containing HTML code, your class should print the headings, one per line and in the order in which they appear in the document. Each heading should be indented as follows: an h1 heading should have indentation 0, and h2 heading should have indentation 1, etc. Test your implementation using w3c.html.  

>>>  infile  =  open('w3c.html')

>>>  content  =  infile.read()

>>>  infile.close()

>>>  hp  =  HeadingParser()

>>>  hp.feed(content)

- W3C  
- Mission Principles

In [52]:
from html.parser import HTMLParser


class HeadingParser(HTMLParser):
    def __init__(self):
        super().__init__()
        self.indentation_level = 0
        self.ignore_data = False

    def handle_starttag(self, tag, attrs):
        if tag.startswith('h'):
            try:
                self.indentation_level = int(tag[1:]) - 1
            except ValueError:
                self.indentation_level = 0
        elif tag == 'title':
            self.ignore_data = True

    def handle_endtag(self, tag):
        if tag == 'title':
            self.ignore_data = False

    def handle_data(self, data):
        stripped_data = data.strip()
        if stripped_data and not self.ignore_data and self.indentation_level >= 0:
            heading_text = stripped_data.lstrip()
            print(' ' * self.indentation_level + heading_text)

In [53]:
# Test the implementation
infile = open('/content/w3c.html')
content = infile.read()
infile.close()

hp = HeadingParser()
hp.feed(content)

W3C Mission
 Principles


##Question 7

- Implement recursive function webdir() that takes as input: a URL (as a string) and non-negative integers depth and indent. Your function should visit every web page reachable from the starting URL web page in depth clicks or less, and print each web page's URL. As shown below, indentation, specified by indent,  should be used to indicate the depth of a URL.  

>>> webdir('http://reed.cs.depaul.edu/lperkovic/csc242/test1.html' ,  2,  0)

>http://reed.cs.depaul.edu/lperkovic/csc242/test1.html

>http://reed.cs.depaul.edu/lperkovic/csc242/test2.html

>http://reed.cs.depaul.edu/lperkovic/csc242/test4.html

>http://reed.cs.depaul.edu/lperkovic/csc242/test3.html

>http://reed.cs.depaul.edu/lperkovic/csc242/test4.html

In [5]:
import urllib.request
from urllib.parse import urljoin
from urllib.error import HTTPError
from bs4 import BeautifulSoup


def webdir(url, depth, indent):
    if depth < 0:
        return

    try:
        print('Fetching:', url)
        response = urllib.request.urlopen(url)
        soup = BeautifulSoup(response.read(), 'html.parser')
        print(' ' * indent + url)

        if depth == 0:
            return

        links = soup.find_all('a')
        for link in links:
            href = link.get('href')
            absolute_url = urljoin(url, href)
            try:
                webdir(absolute_url, depth - 1, indent + 4)
            except HTTPError:
                continue

    except (urllib.error.URLError, ValueError) as e:
        print('Error:', e)
        return


In [6]:
webdir('http://reed.cs.depaul.edu/lperkovic/csc242/test1.html', 2, 0)

Fetching: http://reed.cs.depaul.edu/lperkovic/csc242/test1.html
Error: HTTP Error 404: 


###*While trying to fetch the links, I realised that its giving me an **Error 404**. Thus, I tried to check the code with a common public link and the code works well.*

In [8]:
webdir('https://www.python.org/', 1, 0)

Fetching: https://www.python.org/
https://www.python.org/
Fetching: https://www.python.org/#content
    https://www.python.org/#content
Fetching: https://www.python.org/#python-network
    https://www.python.org/#python-network
Fetching: https://www.python.org/
    https://www.python.org/
Fetching: https://www.python.org/psf/
    https://www.python.org/psf/
Fetching: https://docs.python.org
    https://docs.python.org
Fetching: https://pypi.org/
    https://pypi.org/
Fetching: https://www.python.org/jobs/
    https://www.python.org/jobs/
Fetching: https://www.python.org/community-landing/
    https://www.python.org/community-landing/
Fetching: https://www.python.org/#top
    https://www.python.org/#top
Fetching: https://www.python.org/
    https://www.python.org/
Fetching: https://psfmember.org/civicrm/contribute/transact?reset=1&id=2
    https://psfmember.org/civicrm/contribute/transact?reset=1&id=2
Fetching: https://www.python.org/#site-map
    https://www.python.org/#site-map
Fetchi

##Question 8

a) SQL query to return all the temperature data:


```
SELECT Temperature
FROM table_name;

```

b) SQL query to return all the cities without repetition:


```
SELECT DISTINCT City
FROM table_name;

```

c) SQL query to return all the records for India:


```
SELECT *
FROM table_name
WHERE Country = 'India';

```

d) SQL query to return all the Fall records:


```
SELECT *
FROM table_name
WHERE Season = 'Fall';

```

e) SQL query to return the city, country, and season for which the average rainfall is between 200 and 400 millimeters:


```
SELECT City, Country, Season
FROM table_name
WHERE Rainfall >= 200 AND Rainfall <= 400;

```

f) SQL query to return the city and country for which the average Fall temperature is above 20 degrees, in increasing temperature order:


```
SELECT City, Country
FROM table_name
WHERE Season = 'Fall'
GROUP BY City, Country
HAVING AVG(Temperature) > 20
ORDER BY AVG(Temperature) ASC;

```

g) SQL query to calculate the total annual rainfall for Cairo:


```
SELECT SUM(Rainfall) AS TotalRainfall
FROM table_name
WHERE City = 'Cairo';

```

h) SQL query to calculate the total rainfall for each season:


```
SELECT Season, SUM(Rainfall) AS TotalRainfall
FROM table_name
GROUP BY Season;

```




##Question 9

Suppose list words is defined as follows:
>>>  words  =  ['The',  'quick',  'brown',  'fox',  'jumps',  'over', 'the',  'lazy',  'dog']

Write list comprehension expressions that use list words and generate the following lists:

a) ['THE',  'QUICK',  'BROWN',  'FOX',  'JUMPS',  'OVER',  'THE', 'LAZY',  'DOG']

b) ['the',  'quick',  'brown',  'fox',  'jumps',  'over',  'the', 'lazy',  'dog']

c) [3,  5,  5,  3,  5,  4,  3,  4,  3] (the list of lengths of words in list words).

d) [['THE',  'the',  3],  ['QUICK',  'quick',  5],  ['BROWN', 'brown',  5],  ['FOX',  'fox',  3],  ['JUMPS',  'jumps',  5], ['OVER',  'over',  4],  ['THE',  'the',  3],  ['LAZY',  'lazy', 4],  ['DOG',  'dog',  3]] (the list containing a list for every word of list words, where each list contains the word in uppercase and lowercase and the length of the word.)

e) ['The',  'quick',  'brown',  'fox',  'jumps',  'over',  'the', 'lazy',  'dog'] (the list of words in list words containing 4 or more characters.)

In [9]:
words = ['The', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog']

#a) List comprehension to generate the list of words in uppercase:
uppercase_words = [word.upper() for word in words]

#b) List comprehension to generate the list of words in lowercase:
lowercase_words = [word.lower() for word in words]

#c) List comprehension to generate the list of lengths of words:
word_lengths = [len(word) for word in words]

#d) List comprehension to generate a list of lists containing the word in uppercase, lowercase, and length of the word:
word_info = [[word.upper(), word.lower(), len(word)] for word in words]

#e) List comprehension to generate the list of words containing 4 or more characters:
words_4_or_more_chars = [word for word in words if len(word) >= 4]


In [10]:
print(uppercase_words)
print(lowercase_words)
print(word_info)
print(word_lengths)
print(words_4_or_more_chars)

['THE', 'QUICK', 'BROWN', 'FOX', 'JUMPS', 'OVER', 'THE', 'LAZY', 'DOG']
['the', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog']
[['THE', 'the', 3], ['QUICK', 'quick', 5], ['BROWN', 'brown', 5], ['FOX', 'fox', 3], ['JUMPS', 'jumps', 5], ['OVER', 'over', 4], ['THE', 'the', 3], ['LAZY', 'lazy', 4], ['DOG', 'dog', 3]]
[3, 5, 5, 3, 5, 4, 3, 4, 3]
['quick', 'brown', 'jumps', 'over', 'lazy']
